### PR TITLE
HSEARCH-3256 Restore support for bypassing bridges in the predicate/sort DSL

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3256

~~This is just a preview. Since it still misses:~~ not anymore!
* ~~Tests~~ done!
* ~~Javadoc~~ done!
* ~~Doc update~~ see [HSEARCH-3270](https://hibernate.atlassian.net/browse/HSEARCH-3270)

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/engine/search/predicate/DslConverter.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.search.predicate;
+
+public enum DslConverter {
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/search/predicate/DslConverter.java]: If this class is not used in any API, I'd rather move it to an SPI package for now. Let's not expose anything to users unnecessarily, that will just prevent us from changing it...

Comment (fax4ever) [reply engine/src/main/java/org/hibernate/search/engine/search/predicate/DslConverter.java]: Good idea!

**Comment by fax4ever** [comment]
Sorry I missed to add the compatibility check.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchFieldPredicateBuilderFactory.java [hunk 2]
```diff
@@ -46,13 +47,13 @@
 	 * @return {@code true} if the given predicate builder factory is DSL-compatible.
 	 * {@code false} otherwise, or when in doubt.
 	 */
-	boolean isDslCompatibleWith(ElasticsearchFieldPredicateBuilderFactory other);
+	boolean isDslCompatibleWith(ElasticsearchFieldPredicateBuilderFactory other, DslConverter dslConverter);
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchFieldPredicateBuilderFactory.java]: You forgot to add the new parameter to the javadoc.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneFieldPredicateBuilderFactory.java [hunk 3]
```diff
@@ -45,13 +46,13 @@
 	 * @return {@code true} if the given predicate builder factory is DSL-compatible.
 	 * {@code false} otherwise, or when in doubt.
 	 */
-	boolean isDslCompatibleWith(LuceneFieldPredicateBuilderFactory other);
+	boolean isDslCompatibleWith(LuceneFieldPredicateBuilderFactory other, DslConverter dslConverter);
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneFieldPredicateBuilderFactory.java]: You forgot to add the new parameter to the javadoc.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/impl/IndexSchemaFieldNodeComponentRetrievalStrategy.java [hunk 4]
```diff
@@ -7,14 +7,15 @@
...
+import org.hibernate.search.engine.search.predicate.spi.DslConverter;
 import org.hibernate.search.util.common.reporting.EventContext;
 import org.hibernate.search.util.common.SearchException;
 
 public interface IndexSchemaFieldNodeComponentRetrievalStrategy<T> {
 
 	T extractComponent(ElasticsearchIndexSchemaFieldNode<?> schemaNode);
 
-	boolean areCompatible(T component1, T component2);
+	boolean areCompatible(T component1, T component2, DslConverter dslConverter);
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/impl/IndexSchemaFieldNodeComponentRetrievalStrategy.java]: Thank you!

**Comment by fax4ever** [comment]
@yrodiere thank you for the review, I agree with the change to the javadocs and I think I've fixed up the things you pointed out.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchSearchPredicateBuilderFactoryImpl.java [hunk 5]
```diff
@@ -15,7 +15,7 @@
 import org.hibernate.search.backend.elasticsearch.search.impl.IndexSchemaFieldNodeComponentRetrievalStrategy;
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchSearchPredicateBuilderFactoryImpl.java]: There's still a typo... I'll fix it.

**Comment by yrodiere** [comment]
Merged, thanks!

**Comment by fax4ever** [comment]
Thanks!

**Pull request merged**